### PR TITLE
Metrics team

### PR DESCRIPTION
**Pull request by CaydenEhrlich**
Implemented all metrics besides accuracy and true positive or true negative rate (which was only updated).

### CODE REVIEW

File: Metrics.py [hunk 1]
```diff
@@ -37,16 +47,323 @@ def truePosOrNeg(self, dataSet, truePosOrNeg):
...
+	'''
+	def truePosOrNegRate(self, matchesLabel, actualLabel):
+		return matchesLabel / actualLabel * 100
+
+	'''
+	Segments the overarching DataSet into smaller (non-overlapping) subsets by protected attribute. Returns a list of these
+	subsets as well as a list of the corresponding protected attribute values.
+		dataSet (DataSet) - the overarching DataSet NOTE WE'RE PROBABLY GETTING RID OF THIS PART WHEN WE MAKE IT AN INSTANCE VARIABLE
+	'''
+	def determineGroups(self, dataSet):
```
Comment (mitchellm7) [inline Metrics.py:73]: I was under the assumption that we were only using binary sensitive attributes? Is there a reason this function accounts for more than two possible protected attribute values?

File: Metrics.py [hunk 2]
```diff
@@ -37,16 +47,323 @@ def truePosOrNeg(self, dataSet, truePosOrNeg):
...
+
+			#resets indices for later indexing
+			newDataSet.dataFrame.reset_index(inplace=True, drop=True)
+
+			organizedDataSetList.append(newDataSet)
+
+		return organizedDataSetList, possibleGroups
+
+	'''
+	Performs a chi square test.
```
Comment (mitchellm7) [inline Metrics.py]: Might be useful to add a brief definition of a chi square test.

File: Metrics.py [hunk 3]
```diff
@@ -37,16 +47,323 @@ def truePosOrNeg(self, dataSet, truePosOrNeg):
...
+		
+	returns: a dictionary where the keys are protected attribute values and the values are the number of positive
+		outcomes that protected attribute group received.
+	'''
+	def countPositiveOutcomes(self, dataSet):
+		dataFrame = dataSet.dataFrame
+		possibleAttributes = dataFrame[dataSet.protectedAttributes[0]].unique()
+
+		posOutcomes0 = 0
+		posOutcomes1 = 0
```
Comment (mitchellm7) [inline Metrics.py:227]: Changing the names of these variables or explaining what they hold in the function comment may be useful for readability.

File: Metrics.py [hunk 4]
```diff
@@ -37,16 +47,323 @@ def truePosOrNeg(self, dataSet, truePosOrNeg):
...
+	returns: a dictionary where the keys are the protected attribute values and the values are the positive outcome rate
+		for that protected attribute value.
+	'''
+	def groupFairness(self, dataSet):
+		dataFrame = dataSet.dataFrame
+		posOutcomes = self.countPositiveOutcomes(dataSet)
+		keys = posOutcomes.keys()
+
+		probabilitiesDict = {}
+		totalsFrame = dataFrame[dataSet.protectedAttributes[0]].value_counts()
```
Comment (mitchellm7) [inline Metrics.py]: What exactly is contained within totalsFrame?

**Review by mitchellm7 — APPROVED** [review]
Made a couple comments. Overall looks great! You still have quite a few #TODO comments though, so don't forget about those.

File: Metrics.py (hunk 1 above)
Comment (CaydenEhrlich) [reply Metrics.py:73]: We thought that we had fixed the protectedAttributes but we messed up. Thanks!

**Pull request merged**